File: emisor-js/capas/aplicacion.js
import readline from "readline/promises";
import { stdin as input, stdout as output } from "process";

const rl = readline.createInterface({ input, output });

/**
 * Pide un texto al usuario.
 * @param {string} prompt
 * @returns {Promise<string>}
 */
export async function solicitarMensaje(prompt) {
  return rl.question(prompt);
}

// Se exporta la misma instancia para que otras capas (ej. ruido.js) pidan
// datos por el MISMO readline en vez de crear una interfaz nueva sobre el
// mismo stdin: tener dos interfaces sobre el mismo stream cuelga la lectura.
export function obtenerInterfazEntrada() {
  return rl;
}

/**
 * Pide el algoritmo de integridad a usar.
 * @returns {Promise<string>} "hamming" | "fletcher" | "crc32"
 */
export async function solicitarAlgoritmo() {
  const resp = await rl.question("Algoritmo de integridad (hamming/fletcher/crc32): ");
  return resp.trim().toLowerCase();
}

/**
 * Muestra un mensaje al usuario. Si error es true, se indica que no fue
 * posible corregir/verificar el mensaje.
 * @param {string} mensaje
 * @param {boolean} error
 */
export function mostrarMensaje(mensaje, error = false) {
  if (error) {
    console.log(">> ERROR: mensaje recibido con errores no corregibles");
  } else {
    console.log(">> " + mensaje);
  }
}

export function cerrarEntrada() {
  rl.close();
}

File: receptor-py/capas/aplicacion.py
def solicitar_mensaje(mensaje):
    """
    Solicita un mensaje al usuario y lo devuelve.

    Args:
        mensaje (str): El mensaje que se mostrará al usuario.

    Returns:
        str: El mensaje ingresado por el usuario.
    """
    return input(mensaje)


def solicitar_algoritmo():
    """
    Pide al usuario el algoritmo de integridad a usar.

    Returns:
        str: "hamming" | "fletcher" | "crc32"
    """
    return input("Algoritmo de integridad (hamming/fletcher/crc32): ").strip().lower()


def mostrar_mensaje(mensaje, error=False):
    """
    Muestra un mensaje al usuario. Si error es True, se indica que no fue
    posible corregir/verificar el mensaje.

    Args:
        mensaje (str): El mensaje que se mostrará al usuario.
        error (bool): Si True, se imprime como error en lugar del mensaje.
    """
    if error:
        print(">> ERROR: mensaje recibido con errores no corregibles")
    else:
        print(mensaje)
